eox_nelp/programs/api/v1/views.py

```diff
@@ -106,3 +151,66 @@ def post(self, request, course_id):
         update_program_metadata(course_id, serializer.data, request.user)
 
         return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+
+class ProgramsListView(NelpCourseListView):
+    """
+    API view to list courses with program metadata.
+    **Use Cases**
+        - List courses with associated program metadata for request user
+        - List courses filtered by national_id query parameter
+
+    **Example Requests**
+        GET /eox-nelp/api/programs/v1/program-lookup/
+        GET /eox-nelp/api/programs/v1/program-lookup/?national_id=1234567890
+    """
+    authentication_classes = [JwtAuthentication, SessionAuthenticationAllowInactiveUser]
+    permission_classes = [IsAuthenticated, ProgramsLookupPermission]
+
+    @require_national_id_query_param()
+    def get(self, request, *args, **kwargs):
+        """
+        List courses with program metadata for the request user or filtered by national_id.
+        Args:
+            request: HTTP request object
+            *args: Additional arguments
+            **kwargs: Additional keyword arguments
+        Returns:
+            Response with list of courses and their program metadata or error
+        """
+        course_api_list = super().get(request, *args, **kwargs).data["results"]
+
+        program_lookup_list = []
+        for course_api_data in course_api_list:
+            program_lookup = get_program_lookup_representation(course_api_data)
+            program_lookup_serializer = ProgramLookupSerializer(data=program_lookup)
+            if program_lookup_serializer.is_valid():
+                program_lookup_list.append(program_lookup_serializer.data)
+            else:
+                program_lookup_list.append(
+                    {
+                        "error": "Invalid program lookup data",
+                        "details": program_lookup_serializer.errors,
+                        "course_id": course_api_data.get("course_id"),
+                    }
+                )
+
+        return Response(program_lookup_list, status=status.HTTP_200_OK)
+
+    def get_queryset(self):
+        """Override qs to query by national_id if provided.
+        Also filter by enrolled courses only.
+        Returns:
+            List of enrolled courses for the user qs: Depending on performance this list queryset  could be improved to
+            lazy sequence as used in
+            https://github.com/openedx/edx-platform/blob/258f3fc/lms/djangoapps/course_api/api.py#L111
+
+        """
+        user = self.request.user
+        if national_id := self.request.query_params.get("national_id"):
+            user = get_object_or_404(User, extrainfo__national_id=national_id)
+            visible_courses_qs = courses.get_courses(user=user)
+        else:
+            visible_courses_qs = super().get_queryset()
+        enrolled_qs = [course for course in visible_courses_qs if CourseEnrollment.is_enrolled(user, str(course.id))]
+        return enrolled_qs
```

Comment:
I have multiple suggestions here 
1. move the filter logic to  filter_queryset
2. set serializer_class attribute with ProgramLookupSerializer
3. DONT filter the data when the national id is not provided 
4. Implement the programs logic here

```suggestion
        visible_courses = super().get_queryset()  # get visible courses for the requester 
        course_list = CourseDetailSerializer(visible_courses, many=True).data # Serialize data with eox_nelp.edxapp_wrapper.course_api.CourseDetailSerializer
         program_lookup_list = []
 
        for course_data in course_list:
            program_lookup_list.append(get_program_lookup_representation(course_data))
        
        return program_lookup_list ## or return the lazy sequence that mentioned 
```